src/Lykke.blue.Service.ReferralLinks.AzureRepositories/ReferralLink/ReferralLinkRepository.cs

```diff
@@ -85,8 +93,46 @@ public bool IsInvitationLinkForSenderAlreadyCreated(string senderClientId)
                 return x;
             });
 
-            return Mapper.Map<ReferralLinkDto>(result);
+            return result;
         }
 
+        /// <summary>
+        /// Ensures reflink state is not stale at the moment of update. Throws excpetion if it does.
+        /// Relies on ETag checks, performed by both AzureTableStorage db and MergeAsync itself. 
+        /// MergeAsync will return null if recored is stale, due to the Etag check in the injected lamda.
+        /// Code is also thread safe. 
+        /// </summary>
+        /// <param name="referralLink"></param>
+        /// <returns></returns>
+        public async Task<IReferralLink> UpdateAsyncWithETagCheck(IReferralLink referralLink)
+        {
+            await SemaphoreSlim.WaitAsync();
+            try
+            {
+                var result = await _referralLinkTable.MergeAsync(GetPartitionKey(), GetRowKey(referralLink.Id), currentDbRecord =>
```

Comment:
why we here merge entity. Maybe we can replase record by this key, or we try update not all property?

Reply:
The main reason we use MergeAsync is because its implementation in the Lykke.AzureStorage library includes an ETag (hidden time-stamp field used by Azure Table Storage to ensure optimistic concurrency).  And yes, we use it to only update properties (usually just one - the State) of the entity. In fact,  ReplaceAsync is also ETag - aware, and can be used instead of MergeAsync, as the entity is read from db before being updated, so should be safe to entirely replace it on update. I just ended up using MergeAsync.